scripts/un/sdg/geography.py

```diff
@@ -0,0 +1,325 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+'''Generates geographies for UN places.
+
+Produces:
+* un_places.mcf (place definitions)
+* un_containment.mcf (place containment triples)
+* place_mappings.csv (SDG code -> dcid)
+
+Usage: python3 geography.py
+'''
+import collections
+import csv
+import json
+import os
+
+# Output folder.
+FOLDER = 'geography'
+
+PLACE_TEMPLATE = '''
+Node: dcid:{dcid}
+typeOf: dcs:{type}
+name: "{name}"
+unDataCode: "{code}"
+unDataLabel: "{label}"
+'''
+CONTAINMENT_TEMPLATE = '''
+Node: dcid:{dcid}
+typeOf: dcid:{type}{containment}
+'''
+
+# Curated map of SDG GEOGRAPHY_CODE to UN data code.
+FIXED = {
+    'africa': '2',
+    'undata-geo/G99999999': '952',
+}
+
+
+def get_sdg2type(file):
+    '''Produces map of SDG code -> SDG type.
+
+    Args:
+        file: Input file path.
+
+    Returns:
+        Map of SDG code -> SDG type.
+    '''
+    sdg2type = {}
+    with open(file) as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            sdg2type[row['GEOGRAPHY_CODE']] = row['GEOGRAPHY_TYPE']
+    return sdg2type
+
+
+def get_sdg_un_maps(file):
+    '''Produces maps of UN code -> SDG code & SDG code -> UN code.
+
+    Args:
+        file: Input file path.
+
+    Returns: 
+        - Map of UN code -> SDG code.
+        - Map of SDG code -> UN code.
+    '''
+    un2sdg = {}  # Map of UN code -> SDG code.
+    sdg2un = {}  # Map of SDG code -> UN code.
+    with open(file, encoding='utf-8-sig') as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            subject = row['subject_id']  # UN code.
+            object = row['object_id'].removeprefix('sdg-geo:')  # SDG code.
+            if not subject or not object:
+                continue
+            un2sdg[subject] = object
+            sdg2un[object] = subject
+    return un2sdg, sdg2un
+
+
+def get_un2dc(file):
+    '''Produces map of UN code -> curated (dcid, DC type, DC name).
+
+    Args:
+        file: Input file path.
+
+    Returns:
+        Map of UN code -> curated (dcid, DC type, DC name).
+    '''
+    un2dc = {}
+    with open(file) as f:
```

Comment:
Is there some place that all the input structures are documented that we can link to?

Reply:
updated the structures and added comment